### ISSUE TITLE
Threading performance issue on Windows

### ISSUE THREAD

**Issue by mseminatore** [xianyi/OpenBLAS#4063]
Using 0.3.23 pre-built x64 binary DLL. Also tested against locally built (via cmake/ninja/clang) DLL and static lib sync'd both to head of `develop` branch and to the `v0.3.23` label. Code runs very slowly on Windows 10/11 tested against Sandybridge, Haswell and Zen2/threadripper. Tested against 4 cores, 8 cores, 24 cores and 128 cores.

Noted that on several machines with more than 8 cores only 1 core appears to be busy. On the Sandybridge machine 8 core all cores seem to be taking work.

The same code (both mine and OpenBLAS) runs dramatically (10x) faster on both M1 iMac as well as Intel MacPro. Could there be an issue with either the Windows build config or threading?

Also, if it set OPENBLAS_NUM_THREAD=1 on the Windows machines I see up to a 10x speedup.

My workload is a vectorized C NN library with the same NN structure and dataset as the TensorFlow MNIST Fashion tutorial. Primary CBLAS functions used are: SGER and SGEMV (over 65% execution time spent in those functions). Also used with much lower frequency:  SAXPY and SAXPBY.

Following two runs were on the same machine. In this case a Windows 11 Surface Laptop but similar results seen across many different Windows 10/11 desktop machines.

OpenBLAS 0.3.23 DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=64
      CPU uArch: Haswell
  Cores/Threads: 8/8

Training ANN
------------
  Network shape: 784-128-10
      Optimizer: Adaptive SGD
  Loss function: Categorical cross-entropy
Mini-batch size: 8
  Training size: 60000 rows

Epoch 1/5
[====================] - loss: 0.31 - LR: 0.05
Epoch 2/5
[====================] - loss: 0.68 - LR: 0.047
Epoch 3/5
[====================] - loss: 0.12 - LR: 0.011
Epoch 4/5
[====================] - loss: 0.38 - LR: 0.016
Epoch 5/5
[====================] - loss: 0.27 - LR: 0.0037

Training time: 140.000000 seconds, 0.466667 ms/step
Test accuracy: 86.97%

C:\dev\NN>set OPENBLAS_NUM_THREADS=1

C:\dev\NN>ann
OpenBLAS 0.3.23 DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=64
      CPU uArch: Haswell
  Cores/Threads: 8/1

Training ANN
------------
  Network shape: 784-128-10
      Optimizer: Adaptive SGD
  Loss function: Categorical cross-entropy
Mini-batch size: 8
  Training size: 60000 rows

Epoch 1/5
[====================] - loss: 0.31 - LR: 0.05
Epoch 2/5
[====================] - loss: 0.68 - LR: 0.047
Epoch 3/5
[====================] - loss: 0.12 - LR: 0.011
Epoch 4/5
[====================] - loss: 0.38 - LR: 0.016
Epoch 5/5
[====================] - loss: 0.27 - LR: 0.0037

Training time: 13.000000 seconds, 0.043333 ms/step
Test accuracy: 86.97%

**Comment by mseminatore** [comment]
I've taken an Xperf capture and in WPA I can see that the vast majority of time is spent in context switching. As a result, more than 57% of the time is spent in the idle task.

Something looks very wrong with how threads are interacting. I am seeing a context switch rate of nearly 30K/s. Typically that number wants to be <100/s up to maybe 1000/s. You lose about 8 - 12 uS per context switch so that is a lot of lost performance.

If someone can point me to where in the code this is handled, I can maybe take a closer look.

**Comment by brada4** [comment]
Are 2 threads slow too?

**Comment by mseminatore** [comment]
Yes. Two threads were also slow. I will double check. Sent from my iPhone XOn May 26, 2023, at 5:11 PM, Andrew ***@***.***> wrote:﻿
Are 2 threads slow too?

**Comment by brada4** [comment]
Typical suspect is too early start of threading.
Examine files in interface/ 
SGER (SMP is disabled) and SGEMV (SMP - gemv.c) 
SAXPY (SMP - axpy.c) and SAXPBY (no SMP)

Add `#undef SMP` as first line in SMP-enabled files to confirm they are the culprit.

If yes, check which of 2 had bigger damage, and go doubling threading threshold and measuring if any improvement is reached.

**Comment by martin-frbg** [comment]
interface/gemv.c line 229, multiply by a factor of about 10 and recompile everything. (i.e. multithreading starts at too small matrix sizes, could be thread switching/spinning is extra expensive on Windows.Am working on this but family emergencies keep getting in the way still)

**Comment by brada4** [comment]
Z13 threshold in AXPY seems more modern too.

**Comment by mseminatore** [comment]
> interface/gemv.c line 229, multiply by a factor of about 10 and recompile everything. (i.e. multithreading starts at too small matrix sizes, could be thread switching/spinning is extra expensive on Windows.Am working on this but family emergencies keep getting in the way still)

Thank you Martin, I will explore what you brada4 have suggested and report back what I learn. More than happy to help sort the problem, I am new to the codebase and so need pointers.

And I completely understand and can relate to the challenge of family emergencies. Please take care.

**Comment by mseminatore** [comment]
> Are 2 threads slow too?

Confirmed that 2 threads are slower than one. Almost twice as slow on my Sandybridge machine. I am trying the `#undef SMP` option, first with gemv then with others to see how that goes. I will report back what I find.

For some added context, with the NN structure of 784-128-10 the first gemv call will be a matrix of 784x128 = 100K with a vector size of 128. The SGER calls will similarly be an outer product of 784 x 128.

**Comment by brada4** [comment]
It is only in gemv and axpy.
i.e
first identify that surplus threading hurts function X
then try to tune function X threading threshold.

**Comment by mseminatore** [comment]
Thank you, doing that now. Sent from my iPhone XOn May 27, 2023, at 11:23 AM, Andrew ***@***.***> wrote:﻿
It is only in gemv and axpy.
i.e
first identify that surplus threading hurts function X
then try to tune function X threading threshold.

**Comment by mseminatore** [comment]
> It is only in gemv and axpy. i.e first identify that surplus threading hurts function X then try to tune function X threading threshold.

First update, applying `#undef SMP` to the very top of gemv.c did not make a difference. Same time with 2 threads, almost 2x slower than 1 thread. Checking axpy.c now.

Is axpby.c implemented in terms of axpy.c so that I don't need to check that file as well?

**Comment by brada4** [comment]
axpby has SMP section disabled, you need to check axpy only.
probably sweep your code with `grep -r "cblas_` if anything else is called.

**Comment by mseminatore** [comment]
Adding `#undef SMP` at the very top of gemv.c and saxpy.c did not change the performance for 2 threads. Any other suggestions?

Can you confirm that the undef should happen before including "common.h", I realized that maybe you meant at the top but after the includes.

The complete list of functions I call are:

- cblas_sgemv
- cblas_sger
- cblas_saxpy
- cblas_saxpby
- cblas_sscal
- cblas_sasum

**Comment by martin-frbg** [comment]
I _think_ it should not matter whether you put the undef before or after the includes - the idea is simply to force the function to run single-threaded no matter the matrix size (which may be too heavy-handed if there are some calls that would benefit from multithreading).
Another thing to investigate - _if_ this turns out to be a Windows-specific issue, despite or on top of the known performance problem (#2846) in GEMV - could be the implementation of the macro YIELDING in common.h
https://github.com/xianyi/OpenBLAS/blob/df88536d1c382886f5456970607f2045dd35a4fe/common.h#L349-L355
which may be causing unnecessary overhead. Maybe redefining this as empty, or as some form of `pause` instruction would have an effect.

**Comment by mseminatore** [comment]
Thank you. I tried with the `#undef SMP` before and after in both `sger` and `gemv`. No noticeable effect, if anything it might have gotten slightly slower by about 1s for two threads. 

I will look into the Yield macro next.

**Comment by WilliamTambellini** [comment]
@mseminatore  keep an eye on what MS calls 'quantum':
https://learn.microsoft.com/en-us/windows-hardware/test/wpt/cpu-analysis
and also thread affinity:
https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setthreadaffinitymask

**Comment by mseminatore** [comment]
@martin-frbg, @brada4 Ok after looking at the xperf captures in WPA it appears the issue is not with extra threads Yielding efficiently (I tried every variation of nop, pause, Sleep(0), SwitchToProcessor(), etc. I could think of.

Instead, most of the threads are idling efficiently. The main thread seems to be creating and deleting critical sections at a very high frequency as well as spending as much as 50% of its time waiting. Looking into `blas_server_win32.c` as the possible source. 
...
At the top of the offending call stacks is exec_blas(). Still trying to unravel the threading pattern, but is there a reason why the code wants to create/delete an event and a crit section every time exec_blas is called rather than create/cache the values? On Windows it is very expensive to create/delete kernel objects like crit sections, events, etc.

**Comment by mseminatore** [comment]
What I believe is the problematic code:

```C
int exec_blas_async(BLASLONG pos, blas_queue_t *queue){

#if defined(SMP_SERVER)
  // Handle lazy re-init of the thread-pool after a POSIX fork
  // on Cygwin or as delayed init when a static library	is used
  if (unlikely(blas_server_avail == 0)) blas_thread_init();
#endif

  blas_queue_t *current;

  current = queue;

  while (current) {
    InitializeCriticalSection(&current -> lock); // why create a crit section here for each queue entry?
    current -> finish = CreateEvent(NULL, FALSE, FALSE, NULL); // why create a new event here for each queue entry?
    current -> position = pos;

#ifdef CONSISTENT_FPCSR
    __asm__ __volatile__ ("fnstcw %0"  : "=m" (current -> x87_mode));
    __asm__ __volatile__ ("stmxcsr %0" : "=m" (current -> sse_mode));
#endif

    current = current -> next;
    pos ++;
  }

  EnterCriticalSection(&pool.lock);

  if (pool.queue) {
    current = pool.queue;
    while (current -> next) current = current -> next;
    current -> next = queue;
  } else {
    pool.queue = queue;
  }

  LeaveCriticalSection(&pool.lock);

  SetEvent(pool.filled);

  return 0;
}

int exec_blas_async_wait(BLASLONG num, blas_queue_t *queue){

#ifdef SMP_DEBUG
    fprintf(STDERR, "Synchronization Waiting.\n");
#endif

    while (num){
#ifdef SMP_DEBUG
    fprintf(STDERR, "Waiting Queue ..\n");
#endif

      WaitForSingleObject(queue->finish, INFINITE);

      CloseHandle(queue->finish); // we are deleting the finish event here, and creating a new one later on!?
      DeleteCriticalSection(&queue -> lock); // deleting a crit sectino here and and creating a new one later on?!

      queue = queue -> next;
      num --;
    }

#ifdef SMP_DEBUG
    fprintf(STDERR, "Completely Done.\n\n");
#endif

  return 0;
}
```

**Comment by martin-frbg** [comment]
I suspect this could be simply down to a Unix developer googling what the Microsoft equivalent of a mutex is...

**Comment by brada4** [comment]
Lock whole queue to reduce number of syscalls? cf
https://github.com/xianyi/OpenBLAS/pull/1785/files

**Comment by martin-frbg** [comment]
BTW `git blame` reveals this to be original GotoBLAS code from about 15 years ago-

**Comment by mseminatore** [comment]
No problem! I totally understand that CPU's and OS's have evolved over the years. And we are solving much larger problems on desktop machine than ever before. So the parameters have changed.

I am more than happy to try to help out and contribute a solution if I can. I'll play around with a few ideas and report back. 

It's slow going as I'm trying to learn the code and unravel the call trees. For example, I can't tell yet if the queued work items are created or if they are re-used (e.g. queue associated with a function/kernel). If they are re-used then the fix might be as simple as create the Event and CritSec if and only if it has not yet been created. If they are ephemeral, then I'll explore caching a pool of them so that they can be assigned as needed and then returned to the pool when done instead of destroying them repeatedly.

Any pointers to save me some time digesting the codebase are welcome.

**Comment by mseminatore** [comment]
@martin-frbg @brada4 I have an updated version of `blas_server_win32.c` which updates the threading model to be a bit more modern and performant on Windows. Basically I've re-worked the code to remove all Events, Critical Sections and Waits in favor of a lockless design. Which appears to be how the *nix `blas_server.c` version works. In testing, this seems to dramatically improve the performance with increasing threads by reducing the context switch overhead. At the cost of spinning some worker threads.

More importantly it passes the tests run by the cmake build.

Before I send over a PR, I've had to make a few assumptions and I want to check those here:

1. For Windows, `blas_exec_async()` looks like it was written to be re-entrant. Meaning, we can add a new work queue to an existing work queue. However, it does not appear that is the case for any calling function. And it looks to me like the *nix version does not support that. Can you confirm that I am correct?
2. It looks like `blas_thread_shutdown()` is never called. Is that correct?
3. Is there a minimum version of Windows that is supported? The docs do not make that clear. I'm only testing 10/11.
4. I assume that we cannot require stdatomic.h?
5. Is 32b Windows still supported?
6. Are there other tests to run besides the ones included in the cmake build?

**Comment by brada4** [comment]
6/ Please see if your initial tests get any improvements
5/ Ancient applications come in 32bit versions, you must be able to replace blas.dll included there for foreseeable future
4/ You can, wrapped in ifdef-s
3/ Should not break in foreseeable future (5), those Win2000 applications will not magically start supporting Windows 12 and latest visual studio runtimes.
1-2/ It is a port from pthread-s, if you see better way to organize threading on Windows you are more than welcome.

**Comment by mseminatore** [comment]
@brada4 Thank you, that is helpful information.

6) As mentioned above, my tests show a dramatic improvement in throughput. On Windows, the way that threading was structured, well over 30% of the execution time was spent with threads sleeping/waiting. For example, worker threads were kicked off serially by the main thread setting an Event. That produced a waterfall cascade of worker thread starts with a potential context switch on every `SetEvent()` call. And every worker thread completing marked its completion via a `SetEvent()`, another context switch. And worse, the worker threads created/deleted a `CriticalSection` and `Event` **every time a BLAS function was called**. For my NN test case that was **many tens of Millions** of kernel objects created/destroyed. Every call into an OS function is a potential context switch.

I was mainly asking about compliance tests to ensure that I haven't broken anything. For now, I will have to assume that if it passes my use cases and the suite of tests provided in the repo that all is well.

4) If there has to be a code path that doesn't use stdatomic.h then there seems to be little point in using atomic constructs, that's just more code paths to test and maintain given that the alternate code paths must still be written to behave atomically for the algorithms to work. It seems better to have a single code path that works since that reduces the overall testing permutations. Please let me know if I am misunderstanding.

This code in the *nix `blas_server.c` code path is an example. If the non C11 path is ever correct, why not use it in all cases? If it is ever not correct for a compiler/processor then isn't that a significant issue? Specifically, wouldn't any compiler that doesn't support C11 and doesn't handle volatile as expected (e.g. non-ISO conformant) might subtly misbehave on that code? I guess if the tests would reliably catch that then all is well.

```c
#ifdef HAVE_C11
#define	atomic_load_queue(p)		__atomic_load_n(p, __ATOMIC_RELAXED)
#define	atomic_store_queue(p, v)	__atomic_store_n(p, v, __ATOMIC_RELAXED)
#else
#define	atomic_load_queue(p)		(blas_queue_t*)(*(volatile blas_queue_t**)(p))
#define	atomic_store_queue(p, v)	(*(volatile blas_queue_t* volatile*)(p) = (v))
#endif
`

**Comment by brada4** [comment]
4. Think back to RHEL5 VS2010 WIN2000, guard old codepath wihout any changes with ifdef-s. It used to work well on w7/2008 even after spectre series. You still can buy support for those.

**Comment by martin-frbg** [comment]
It is hard to comment without knowing what your code looks like, but 
(1) the history of OpenBLAS/GotoBLAS goes back a bit before the creation of C11, and there may still be users who are unwilling or unable to upgrade old compilers (or sometimes even old hardware). So any C11-idioms in OpenBLAS are a
relatively recent improvement that comes with backward compatibility. (Typically in the simple form of keeping the less efficient or even less reliable code path available behind an `#ifndef HAVE_C11` - if this appears too cumbersome in your case, it would also be possible to handle this at the file level, with conditional compilation of a suitably named "old" or "new" version of blas_server_win32.c).  
(2) In a similar fashion, the included tests have grown in number with demand, and can/will continue to do so. The ones in "test" and "ctest" were simply lifted from the reference implementation of BLAS, and we also carry the LAPACK testsuite. I _think_ these should be able to catch at least the most glaring bugs with multithreading, particularly if run on a larger hardware. There are also the DGEMM and DGEMV tests in cpp_thread_test that are not built by default (as they require a C++ compiler), but can be enabled with `-DCPP_THREAD_SAFETY_TEST=ON` and `-DCPP_THREAD_SAFETY_GEMV=ON` .

**Comment by mseminatore** [comment]
Thank you. I am getting close to having a PR ready for review. I'm avoiding adding explicit usage of stdatomic.h for now, it's not required for the solution that I have working.

I think I've found a bounds issue in `goto_set_num_threads()` so I will probably send that PR over first since it's a 1-line fix. I'll open a separate issue for that.

**Comment by mseminatore** [comment]
PR sent #4103 

My use case NN library is now publicly available [here](https://github.com/mseminatore/ann).

In my tests on 8-core Sandybridge, 8-core Haswell and 64-core Zen Threadripper running Windows 10/11 this code is faster than the previous versions.

Here are my results from the 8-core Sandybridge running 5 epochs, 60k rows (= 300,000 training iterations) training NN against MNIST Fashion dataset with network shape of 784-128-10 (roughly 100k weights).

**3.23 develop branch OpenBLAS**

1 thread | 8 threads
----------|-----------
15 seconds | 51 seconds

**This PR**

1 thread | 8 threads
----------|-----------
16 seconds | 19 seconds
